compiler/rustc_expand/src/mbe/metavar_expr.rs

```diff
@@ -0,0 +1,285 @@
+
+        Some(MetaVarExpr::Length(depth.unwrap_or(0)))
+    }
+}
+
+/// A meta-variable expression argument.
+enum Arg {
+    /// An identifer argument.
+    Ident(Ident),
+
+    /// A literal argument.
+    Literal(token::Lit, Span),
+}
+
+impl Arg {
+    /// The span that contains the argument.
+    fn span(&self) -> Span {
+        match *self {
+            Arg::Ident(ident) => ident.span.clone(),
+            Arg::Literal(_, span) => span.clone(),
+        }
+    }
+
+    /// Attempt to parse the argument as an identifier.
+    fn parse_ident(self, sess: &ParseSess) -> Option<Ident> {
+        match self {
+            Arg::Ident(ident) => Some(ident),
+            Arg::Literal(_, span) => {
+                let msg = "expected identifier";
+                sess.span_diagnostic.span_err(span, msg);
+                None
+            }
+        }
+    }
+
+    /// Attempt to parse the identifier as a literal and convert it
+    /// to a parsed type.
+    fn parse_literal<T: FromStr>(self, sess: &ParseSess) -> Option<T>
+    where
+        <T as FromStr>::Err: Display,
+    {
+        match self {
+            Arg::Literal(lit, span) => {
+                if lit.suffix.is_some() {
+                    let msg = "literal suffixes are not supported in meta-variable expressions";
+                    sess.span_diagnostic.span_err(span, msg);
+                    return None;
+                }
+                match lit.symbol.as_str().parse::<T>() {
+                    Ok(depth) => Some(depth),
+                    Err(e) => {
+                        sess.span_diagnostic.span_err(span, &format!("failed to parse meta-variable expression argument: {}", e));
+                        return None;
+                    }
+                }
+            }
+            Arg::Ident(ident) => {
+                let msg = "expected literal";
+                sess.span_diagnostic.span_err(ident.span, msg);
+                return None;
+            }
+        }
+    }
+}
+
+/// Attempt to parse meta-variable expression arguments.
+fn parse_args(
+    span: Span,
+    mut input: impl Iterator<Item = tokenstream::TokenTree>,
+    sess: &ParseSess,
+) -> Vec<Arg> {
+    let mut result = Vec::new();
+    let msg = "expected metavariable expression arguments";
+    let args = match input.next() {
+        Some(tokenstream::TokenTree::Delimited(_, token::Paren, args)) => Some(args),
+        Some(tree) => {
+            sess.span_diagnostic.span_err(tree.span(), &msg);
+            None
+        }
+        None => {
+            sess.span_diagnostic.span_err(span, &msg);
+            None
+        }
+    };
+    if let Some(args) = args {
+        let mut tts = args.trees();
+        while let Some(tt) = tts.next() {
+            match tt {
+                tokenstream::TokenTree::Token(token) if token.is_ident() => {
+                    if let Some(ident) = token.ident() {
+                        result.push(Arg::Ident(ident.0))
+                    }
```

Comment:
Also ignores if not found instead of `token.ident().unwrap()`. IIRC, this uncovered an error that was being silenced